### PR TITLE
Add zap opt out support via thrift annotations

### LINKED ISSUE #371: zap: Opt out sensitive struct fields
MarshalLogObject on struct fields should support opting specific fields of a
struct out of being logged in case they contain sensitive information.

    struct User {
        1: required string username
        2: required string emailAddress (go.log = "false")
    }
    
The annotation name is up for discussion. 

CC @prashantv @akshayjshah @kriskowal

### PR DESCRIPTION
**Pull request by mh-park**
Resolves #371

### CODE REVIEW

File: gen/zap.go [hunk 1]
```diff
@@ -26,6 +26,8 @@ import (
 	"go.uber.org/thriftrw/compile"
 )
 
+const NoZapLabel = "go.nolog"
```
Comment (abhinav) [inline gen/zap.go]: This should have docs on it. The linter would have complained about it but I
just realized that the linter hasn't been running in CI for a while. Fixing.

File: gen/zap.go [hunk 2]
```diff
@@ -26,6 +26,17 @@ import (
 	"go.uber.org/thriftrw/compile"
 )
 
+// NoZapLabel allows opt out of Zap logging for struct fields.
+// Thrift structs will use this annotation to omit it from logging, i.e.
+//
+// struct ZapOptOutStruct {
+// 	1: required string name
+// 	2: required string optout (go.nolog)
```
Comment (abhinav) [inline gen/zap.go]: Indent code samples at least one more space to have them render as code blocks.

File: gen/zap.go [hunk 3]
```diff
@@ -26,6 +26,17 @@ import (
 	"go.uber.org/thriftrw/compile"
 )
 
+// NoZapLabel allows opt out of Zap logging for struct fields.
+// Thrift structs will use this annotation to omit it from logging, i.e.
```
Comment (abhinav) [inline gen/zap.go]: Fields of Thrift structs will use this annotation to opt-out of being logged
when that struct is logged.

File: gen/field.go [hunk 4]
```diff
@@ -91,8 +91,10 @@ func (f fieldGroupGenerator) Generate(g Generator) error {
 		return err
 	}
 
-	if err := f.Zap(g); err != nil {
-		return err
+	if !checkNoZap(g) {
```
Comment (amckinney) [inline gen/field.go]: Reads a little funky with the double negative, but I see that the semantics are taken from other similar helper functions like `checkReservedIdentifier`.

File: gen/zap_test.go [hunk 5]
```diff
@@ -476,6 +476,20 @@ func TestCrazyTownZapLogging(t *testing.T) {
 	}
 }
 
+func TestOptOutOfZap(t *testing.T) {
+	// These types are created to ease building map[string]interface{}
```
Comment (amckinney) [inline gen/zap_test.go]: Might not be necessary since it's only used in one place.

**Pull request merged**